File: app/src/main/java/com/yenvth/soilDetectionApp/map/MapAdapter.java
package com.yenvth.soilDetectionApp.map;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.makeramen.roundedimageview.RoundedImageView;
import com.yenvth.soilDetectionApp.R;
import com.yenvth.soilDetectionApp.models.SoilModel;
import com.yenvth.soilDetectionApp.utils.Constant;

import java.util.ArrayList;

public class MapAdapter extends RecyclerView.Adapter<MapAdapter.MyHolder> {
    private Context mContext;
    private ArrayList<SoilModel> dataList;
    private OnSoilMapItemClickListener listener;

    public MapAdapter(Context mContext, ArrayList<SoilModel> dataList, OnSoilMapItemClickListener listener) {
        this.mContext = mContext;
        this.dataList = dataList;
        this.listener = listener;
    }

    public interface OnSoilMapItemClickListener {
        void onSoilMapClickListener(SoilModel soilModel);
    }

    @NonNull
    @Override
    public MyHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(mContext).inflate(R.layout.item_soil_map, parent, false);
        return new MyHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull MyHolder holder, int pos) {
        SoilModel soilModel = dataList.get(pos);
        holder.itemView.setOnClickListener(view -> {
            if (listener != null) {
                listener.onSoilMapClickListener(soilModel);
            }
        });
        holder.lnImage.setBackgroundColor(mContext.getResources().getColor(Constant.getColorArea(soilModel.getSoilCode().toLowerCase())));
        holder.tvName.setText(soilModel.getNameVi());
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    public class MyHolder extends RecyclerView.ViewHolder {
        private TextView tvName;
        private LinearLayout lnImage;
        public MyHolder(@NonNull View itemView) {
            super(itemView);
            tvName = itemView.findViewById(R.id.tvName);
            lnImage = itemView.findViewById(R.id.lnImage);
        }
    }
}

File: app/src/main/java/com/yenvth/soilDetectionApp/map/MapAdapter.kt
package com.yenvth.soilDetectionApp.map

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.yenvth.soilDetectionApp.R
import com.yenvth.soilDetectionApp.models.SoilModel
import com.yenvth.soilDetectionApp.utils.Constant
import java.util.*

class MapAdapter(
    private val mContext: Context,
    private val listener: OnSoilMapItemClickListener
) : RecyclerView.Adapter<MapAdapter.MyHolder>() {
    private var dataList = emptyList<SoilModel>()

    @SuppressLint("NotifyDataSetChanged")
    fun setData(list: List<SoilModel>) {
        this.dataList = list
        notifyDataSetChanged()
    }

    interface OnSoilMapItemClickListener {
        fun onSoilMapClickListener(soilModel: SoilModel)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder {
        val v = LayoutInflater.from(mContext).inflate(R.layout.item_soil_map, parent, false)
        return MyHolder(v)
    }

    override fun onBindViewHolder(holder: MyHolder, pos: Int) {
        val soilModel = dataList[pos]
        holder.itemView.setOnClickListener {
            listener.onSoilMapClickListener(soilModel)
        }
        holder.lnImage.setBackgroundColor(
            mContext.resources.getColor(
                Constant.getColorArea(
                    soilModel.soilCode?.toLowerCase(Locale.getDefault())
                )
            )
        )
        holder.tvName.text = soilModel.nameVi
    }

    override fun getItemCount(): Int {
        return dataList.size
    }

    inner class MyHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvName: TextView
        val lnImage: LinearLayout

        init {
            tvName = itemView.findViewById(R.id.tvName)
            lnImage = itemView.findViewById(R.id.lnImage)
        }
    }
}